controller/dataplane/konnect_extension_test.go

```diff
@@ -0,0 +1,272 @@
+								Kind:  operatorv1alpha1.DataPlaneKonnectExtensionKind,
+								NamespacedRef: operatorv1alpha1.NamespacedRef{
+									Name: "konnect-ext",
+								},
+							},
+						},
+						Deployment: operatorv1beta1.DataPlaneDeploymentOptions{
+							DeploymentOptions: operatorv1beta1.DeploymentOptions{
+								PodTemplateSpec: &corev1.PodTemplateSpec{
+									Spec: corev1.PodSpec{
+										Containers: []corev1.Container{
+											{
+												Name: "proxy",
+												Env: []corev1.EnvVar{
+													{
+														Name:  "KONG_TEST",
+														Value: "test",
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			konnectExt: &operatorv1alpha1.DataPlaneKonnectExtension{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "konnect-ext",
+					Namespace: "default",
+				},
+				Spec: operatorv1alpha1.DataPlaneKonnectExtensionSpec{
+					AuthConfiguration: operatorv1alpha1.KonnectControlPlaneAPIAuthConfiguration{
+						ClusterCertificateSecretRef: operatorv1alpha1.ClusterCertificateSecretRef{
+							Name: "cluster-cert-secret",
+						},
+					},
+					ControlPlaneRef: configurationv1alpha1.ControlPlaneRef{
+						KonnectID: lo.ToPtr("konnect-id"),
+					},
+					ControlPlaneRegion: "us-west",
+					ServerHostname:     "konnect.example.com",
+				},
+			},
+			expectedError: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			objs := []runtime.Object{tt.dataplane}
+			if tt.konnectExt != nil {
+				objs = append(objs, tt.konnectExt)
+			}
+			cl := fake.NewClientBuilder().WithScheme(s).WithRuntimeObjects(objs...).Build()
+
+			dataplane := tt.dataplane.DeepCopy()
+			err := applyDataPlaneKonnectExtension(context.Background(), cl, dataplane)
+			if tt.expectedError {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				requiredEnv := []corev1.EnvVar{}
+				if tt.dataplane.Spec.Deployment.PodTemplateSpec != nil {
+					if container := k8sutils.GetPodContainerByName(&tt.dataplane.Spec.Deployment.PodTemplateSpec.Spec, consts.DataPlaneProxyContainerName); container != nil {
+						requiredEnv = container.Env
+					}
+				}
+
+				if tt.konnectExt != nil {
+					requiredEnv = append(requiredEnv, getKongInKonnectEnvVars(*tt.konnectExt)...)
+					sort.Sort(k8sutils.SortableEnvVars(requiredEnv))
+					assert.NotNil(t, dataplane.Spec.Deployment.PodTemplateSpec)
+					assert.Equal(t, requiredEnv, dataplane.Spec.Deployment.PodTemplateSpec.Spec.Containers[0].Env)
+				}
+			}
+		})
+	}
+}
+
+func getKongInKonnectEnvVars(konnectExt operatorv1alpha1.DataPlaneKonnectExtension) []corev1.EnvVar {
+	envSet := []corev1.EnvVar{}
+	for k, v := range dputils.KongInKonnectDefaults(dputils.KongInKonnectParams{
+		ControlPlane: *konnectExt.Spec.ControlPlaneRef.KonnectID,
+		Region:       konnectExt.Spec.ControlPlaneRegion,
+		Server:       konnectExt.Spec.ServerHostname,
+	}) {
```

Comment:
Nit: IMHO we could put the return value from `KongInKonnectDefaults` into a var.